fvm/programs/block_programs.go

```diff
@@ -1,300 +1,45 @@
 package programs
 
 import (
-	"fmt"
-	"sync"
-
 	"github.com/onflow/cadence/runtime/common"
 	"github.com/onflow/cadence/runtime/interpreter"
 
 	"github.com/onflow/flow-go/fvm/state"
 )
 
-type invalidatableProgramEntry struct {
-	ProgramEntry // Immutable after initialization.
-
-	isInvalid bool // Guarded by blockPrograms' lock.
-}
-
 // BlockPrograms is a simple fork-aware OCC database for "caching" programs
 // for a particular block.
-//
-// Since programs are derived from external source, the database need not be
-// durable and can be recreated on the fly.
-//
-// Furthermore, because programs are derived data, transaction validation looks
-// a bit unusual when compared with a textbook OCC implementation.  In
-// particular, the transaction's invalidator represents "real" writes to the
-// canonical source, whereas the transaction's readSet/writeSet entries
-// represent "real" reads from the canonical source.
 type BlockPrograms struct {
-	lock     sync.RWMutex
-	programs map[common.Location]*invalidatableProgramEntry
-
-	latestCommitExecutionTime LogicalTime
-
-	invalidators chainedInvalidators // Guarded by lock.
+	*BlockDerivedData[common.AddressLocation, ProgramEntry]
 }
 
-// TransactionPrograms is the scratch space for a single transaction.
+// TransactionPrograms is the scratch space for programs of a single transaction.
 type TransactionPrograms struct {
-	block *BlockPrograms
-
-	// The start time when the snapshot first becomes readable (i.e., the
-	// "snapshotTime - 1"'s transaction committed the snapshot view)
-	snapshotTime LogicalTime
-
-	// The transaction (or script)'s execution start time (aka TxIndex).
-	executionTime LogicalTime
-
-	readSet  map[common.Location]*invalidatableProgramEntry
-	writeSet map[common.Location]ProgramEntry
+	*TransactionDerivedData[common.AddressLocation, ProgramEntry]
 
 	// NOTE: non-address programs are not reusable across transactions, hence
```

Comment:
can you add a TODO for me to move non-address programs back into the environment package? It makes less sense to have this here now that we're generalizing the database.  Each TransactionDerivedData is essentially a database table.